Refactoring + ICE support + Ms pixel buffer support + screen/plate/well hierarchy

Rylern opened the pull request:
Initial refactoring of the extension. I'm creating a pull request now to agree on some design decisions:

* I organized packages by **feature**, not by **layer** (see [this link](http://www.javapractices.com/topic/TopicAction.do?Id=205)). For example, the `OmeroExtension` class installs 3 actions/menus:
  * An action sending annotations
  * A browser menu
  * An action to open a connection manager
 
  Each of this feature has one package dedicated to it, located in the same package as the `OmeroExtension` class. 
  The *common* package is an exception, as it contains functions used by every other parts of the project. I'll see if it's possible to integrate some parts of it in separate feature packages.
  With the Javadoc comments, it should be easy to understand how the project is structured.
* I used FXML files each time it was possible, even for very short UI elements.
* Each time a function takes some time to complete (for example when making HTTP requests), I made the function asynchronous using the Java [CompletableFuture](https://www.baeldung.com/java-completablefuture). The functions are then executed in a random thread.
  * Since asynchronous functions are executed in a random thread, I tried to make all non-UI classes thread-safe, which means every public functions of theses classes should work no matter the thread used to call them. I used the `synchronized` keyword for that.
  * When an asynchronous function is complete and I want to update UI elements, I used `Platform.runLater()` to come back to the UI thread.
  * When a JavaFX property or an ObservableList is modified from a thread, the changes are propagated to any listening class on the same thread, which can be a problem if a UI class listens to a property/list changed from a non-UI thread (because UI elements can only be updated from the UI thread). To solve this problem, I created `someUIElementModel` classes that propagate changes made to a property/list from any thread to the UI thread.
* I changed the HTTP requests to use the new [Java HTTP client](https://www.baeldung.com/java-9-http-client). It natively uses CompletableFutures.
* I used [this convention](https://stackoverflow.com/a/3179221) to name packages.
* The texts displayed on the UI are all in one `string.properties` file.
* Unless explicitly written in the documentation, all public functions never return `null`. When a function is not guaranteed to return a result, the Java `Optional` API is used. 

A few other things:
* The ICE API was added to the project and is used if OMERO Java files are present in the classpath. I tested i with 8-bit, 16-bit images and it seems to be working. However, opening large files makes the viewer very slow, so I'm working on this right now.
* In the browser, a label indicates if raw pixel access is available.
* I didn't implement unit tests yet, so this version is not ready to be released.
* I also used the preferences API to store the server URLs so that the user doesn't have to write them each time the application is started.
* I will add support for screens/wells/plates.
* Every issues in the [issue tab](https://github.com/qupath/qupath-extension-omero/issues) should have been addressed.
* Should we use Java modules?

petebankhead commented:
Thank you for doing all this, and the fantastic explanation of everything! I've started to check out the code in more detail but it might take me a little while.

Quick answer to the questions:

> Wouldn't it better to make QuPath `ImageServers` asynchronous? Opening an OMERO image requires to perform several HTTP requests, which for now make the application unresponsive during a few seconds.

Possibly...

I tentatively think it's better to leave that up to the calling code. One reason is that we want the API simple for people to use `server.readRegion(...)` in scripts (which generally run in a background thread anyway). And we definitely can't use any JavaFX features in `ImageServer` since it should be UI-independent. I feel like making the `ImageServer` asynchronous might add complexity and I'm not sure of the specific benefits.

At least, I think it makes sense to retain the current approach *for now*. But I'm strongly in favor of considering how to rethink the *entire* `ImageServer` concept (alongside considering ImgLib2, multidimensional image 'blocks' rather than 2D 'regions', and perhaps writing an entirely new viewer...). So please do record all ideas for improvements :)

> I saw that OMERO was also providing screens/wells/plates, is that something this extension should support?

Ideally yes! Since I *think* these usually aren't RGB images, the previous extension couldn't support the images anyway - so the lack of this feature didn't matter. But it would be a welcome addition.

Rylern commented:
I created a commit with the ICE support, I updated the text above with the changes

petebankhead commented on src/main/java/qupath/lib/images/servers/omero/common/api/authenticators/gui/GuiAuthenticator.java:
```diff
@@ -0,0 +1,40 @@
+
+    @Override
+    protected PasswordAuthentication getPasswordAuthentication() {
+        AuthenticatorForm authenticatorForm = new AuthenticatorForm(getRequestingPrompt(), getRequestingHost(), ClientsPreferencesManager.getLastUsername());
```
This seems to rely on the garbage collector to get rid of the form (and password).
Consider adding a `clear()` method to `AuthenticatorForm` so that the data can be cleared immediately when it is no longer required.

petebankhead commented on src/main/java/qupath/lib/images/servers/omero/common/api/requests/apis/IViewerApi.java:
```diff
@@ -0,0 +1,109 @@
+        }
+    }
+    private List<String> pathObjectsToString(Collection<PathObject> pathObjects) {
+        Gson gsonTMAs  = new GsonBuilder().registerTypeAdapter(TMACoreObject.class, new Shape.GsonShapeSerializer()).serializeSpecialFloatingPointValues().setLenient().create();
```
Is it necessary to have different builders, or could the same builder with multiple type adapters be reused?

petebankhead reviewed (changes requested):
Thanks Léo! This is an enormous piece of work and looking great.

I haven't been able to work through everything in detail, but I've added some comments.

### Organization & scripting

The 'organize by feature' link makes a strong case. My hesitation is that common commands need to be easily 'scriptable' by a user, and work independently of the UI (which includes avoiding dialogs popping up unexpectedly).

'Sending annotations' is an example where this is relevant: this is something a user might certainly want to do in a script.

Creating some example scripts alongside the development would be useful to explore the requirements here and ensure the API meets the user needs.

Related to this, the package organization seems to be very complicated; the names are long, and sometimes repeated (e.g. `package qupath.lib.images.servers.omero.connectionsmanager.connectionsmanager.connection;`).

If lots of classes are internal, then that's not such an issue - making this modular with a `module-info.java` would help make this clear. But also it would help to simplify the package structure if possible, to avoid such extremely long names.

### Switching raw and web pixel access

It needs to be completely obvious to the user when they are getting raw pixel values, and when they get a JPEG-compressed representation - and we need to ensure behavior is consistent ([see discussion on image.sc](https://forum.image.sc/t/java-gateway-authentication-using-the-json-api/84307/9)).

If both options are available, the user could ideally switch between them in the browser - particularly since web access seems much faster.

I think there's currently a bug here, because when I read a raw image (`160701-Nup107-cell-1-t6.tif` from IDR), it was initially very slow as it requested ~500 MB pixels (partly the fault of QuPath's histogram calculation when an image is first opened, so beyond the scope of the extension).

*However* when I reopened the image later, it was much faster and reported using *OMERO (web)*. Upon closer inspection, I could see that it *was* using the web API, quietly and without warning, and returning JPEG-compressed tiles. But it *looked* ok within QuPath (a 2-channel image) until changing the brightness/contrast to reveal the JPEG artifacts. Furthermore, it broke when attempting to send the region to ImageJ (I guess because it really had a 3-channel image, not a 2-channel one).

It should not be possible to quietly regress to non-raw pixel access, so somehow this needs careful checking within the `ImageServer` implementation.


### QuPath compatibility

I think we should target QuPath v0.5.0 for this release, which will require some changes. The biggest is the introduction of [`qupath-fxtras`](https://github.com/qupath/qupath-fxtras) for general code, including the [`Dialogs`](https://github.com/qupath/qupath-fxtras/blob/07d05843a153d94e4d3200332639b68f3fc74dff/src/main/java/qupath/fx/dialogs/Dialogs.java#L62) class.

[`LocalizedResourceManager`](https://github.com/qupath/qupath-fxtras/blob/07d05843a153d94e4d3200332639b68f3fc74dff/src/main/java/qupath/fx/localization/LocalizedResourceManager.java) may also be useful if you want to be able to respond to changes in locale 'live'.

I think updating the code for v0.5.0 compatibility should be the next step, since then we can address any issues (e.g. speed opening images) that actually originate outside the extension.


### Other (minor) things

* Please leave a blank line after the top class definition and before the first field/method is described, for consistency with other classes (including within the JDK itself)
* The browser gives some info below the thumbnail. It would help to add more dimension info, as seen under the 'Image' tab in QuPath *Dimensions (CZT)* and also an estimate of the uncompressed size. Particularly for raw pixel access this helps determine how long it's likely to take to open the image.
* *Extensions &rarr; OMERO &rarr; Manage server connections* doesn't seem to behave properly when expanding the image lists (i.e. the window doesn't resize, and so information is cropped off). Also, I'm not sure that the connections should be maintained after images are closed - and showing the image IDs (rather than names or thumbnails, for example) is not very easy to interpret.
   * In the same command, if I press *Login* and then *Cancel*, I still see an error message that I couldn't log in.
   * Should it be possible to also *Browse* from that dialog?

Rylern commented on src/main/java/qupath/lib/images/servers/omero/connectionsmanager/connectionsmanager/connection/ConnectionModel.java:
```diff
@@ -0,0 +1,57 @@
+
+import java.net.URI;
+
+/**
```
> It looks like the properties exposed by WebClient are largely read-only, so could the effort to ensure they are updated in the application thread be performed within WebClient itself?

This was the initial behavior, I was using `Platform.runLater()` inside `WebClient`. However, when using the extension with the command line (no GUI), an error occurred because the JavaFX application thread couldn't be find. So I decided to remove all `Platform.runLater()` calls from non UI classes.

> (The alternative I also wonder about is whether WebClient should have any dependency on JavaFX at all... but I guess removing properties there would complicate the code and may not have any real-world advantage)

The JavaFX properties are very useful, and can be used even without UI elements, so I'd be in favor of keeping them here.

Rylern commented on src/main/java/qupath/lib/images/servers/omero/common/api/requests/RequestsHandler.java:
```diff
@@ -0,0 +1,326 @@
+    /**
+     * See {@link qupath.lib.images.servers.omero.common.api.requests.apis.WebGatewayApi#readOneDimensionalTile(Long, TileRequest, int, int, double, boolean)}.
+     */
+    public CompletableFuture<Optional<BufferedImage>> readOneDimensionalTile(
```
> If so, is there a need to distinguish them in different methods?

The URL is a bit different between the two methods.

Rylern commented on src/main/java/qupath/lib/images/servers/omero/common/imagesservers/OmeroImageServer.java:
```diff
@@ -0,0 +1,247 @@
+     * @return whether image servers can read any image (e.g. 64-bits multi-channels images).
+     * If not, only uint8 RGB images can be read.
+     */
+    public static boolean canReadAllImages() {
```
Where do you suggest moving it?

Rylern commented:
I updated the pull requests with the changes you requested. This version relies on the [Bio-Formats pull request](https://github.com/qupath/qupath/pull/1327) (this is why the GitHub checks failed).

> Organization & scripting

I reorganized the package structure. There are now less packages, and every UI elements are in a `gui` package.

Do you have ideas of example scripts I should include?

> Switching raw and web pixel access

I changed the way of accessing pixel values. I created the `PixelAPI` interface, which provides information (e.g. supported images) about a pixel API (Ice or web for now). A `PixelAPI` is created by a `WebClient`.

The user can choose an available pixel API in the browser before opening an image. Every OMERO server supports the web API, but not all support the Ice API (for example servers that don't require authentication like IDR).

The default pixel API is the one that provides access to raw pixel values (which is only Ice for now), or the web API if none is available. This means that if a user:
* Browses an OMERO server that supports both Ice and web.
* Opens an image with the **web** API, and saves it.
* Closes QuPath, and later opens the saved image without reselecting the web API.

then the image will be opened with the **Ice** API. Is that a problem?

If the user tries to open an image with an unavailable API, the image server won't be created (which will show an error dialog), so the bug you mentioned shouldn't appear anymore.

`PixelAPI` is used to retrieve information about a pixel API, but can't actually read pixel values. For that, I created the `PixelAPIReader` interface. I separated `PixelAPI` and `PixelAPIReader` because there is only one instance of `PixelAPI` per OMERO server, while there can be several instances of `PixelAPIReader` per OMERO server (if multiple images are opened at the same time for example).

> The browser gives some info below the thumbnail. It would help to add more dimension info, as seen under the 'Image' tab in QuPath Dimensions (CZT) and also an estimate of the uncompressed size. Particularly for raw pixel access this helps determine how long it's likely to take to open the image.

The number of channels, z-slices, and timepoints are indicated, but I didn't find a way to get the uncompressed size from the web API.

petebankhead commented:
Thanks, I'll read properly when I have more time - for now I'll just try to answer the questions.

> I reorganized the package structure. There are now less packages, and every UI elements are in a `gui` package.

Sounds good!

> Do you have ideas of example scripts I should include?

I guess the main one would be to open an image, access pixels or metadata, and close the image.

This may require login (ideally not requiring the password to be in the script).

Also, it should be possible to run a script over OMERO images from a command line, e.g. see https://qupath.readthedocs.io/en/0.4/docs/advanced/command_line.html#subcommands

The following is a rough example - even just printing the image data should demonstrate that the image can be opened.

```
QuPath script -p myproject/project.qpproj -c "println getCurrentImageData()"
```

One might also want to connect to a server and list the images contained, selecting which to open that way.

In the future, there will likely be a need for other things (e.g. requesting or sending objects). Basically, if it can be done through the UI then there is a strong chance a user will want to script it some day...

> then the image will be opened with the Ice API. Is that a problem?

Yes, it's important that we try to ensure the user always receives the same pixels for the same image in their project. If they request the web version, they should *always* receive that - because it will be JPEG-compressed, and therefore have different pixel values.

If it switches to use ICE, then the results of tasks like cell detection will be different (and also the performance will be different).

The choice of server might be encoded in the string args that are associated with `ImageServer` / `ServerBuilder`.

Switching between implementations that give *raw* pixel access may be permitted.

> The number of channels, z-slices, and timepoints are indicated, but I didn't find a way to get the uncompressed size from the web API.

You can calculate this by multiplying all the other dimensions, and then multiplying the result by the bytes per pixel - see
[here](https://github.com/qupath/qupath/blob/82165d4d7089d840f2adf7b502eab37055f1eb45/qupath-gui-fx/src/main/java/qupath/lib/gui/panes/ImageDetailsPane.java#L816) for an example.

Rylern commented:
> The choice of server might be encoded in the string args that are associated with ImageServer / ServerBuilder.

Yes, when an `OmeroImageServerBuilder` is creating an `OmeroImageServer`, I can check for the existence of a specific value in `args` indicating the pixel API to use.

However, how can I set this value? The first time an image is opened, I'd like to add the value of the chosen pixel API to `args`, so that following openings of the image use the same pixel API.

petebankhead commented on build.gradle:
```diff
@@ -1,61 +1,104 @@
-  // To create a shadow/fat jar, including dependencies
-  id 'com.github.johnrengelman.shadow' version '7.0.0'
-  // Reduce JavaCPP dependencies
-  id 'org.bytedeco.gradle-javacpp-platform' version '1.5.7'
```
I think the JavaCPP plugin is still needed, otherwise during builds it will try to get native libraries for all platforms (I don't know why, but calling `./gradlew javadoc` here prompted a lot of unnecessary files to be requested).

You can see how it is added at https://github.com/qupath/qupath-extension-djl/blob/main/build.gradle
with the version in `gradle.settings`: https://github.com/qupath/qupath-extension-djl/blob/main/settings.gradle

Splitting the plugin and version across `build` and `settings` seems necessary for `./gradlew build` and running through an IDE to *both* work simultaneously.
